hasher-matcher-actioner/hmalib/lambdas/pdq/pdq_indexer.py

```diff
@@ -75,33 +126,33 @@ def lambda_handler(event, context):
 
     logger.info("Retreiving PDQ Data from S3")
 
-    with metrics.timer(metrics.names.pdq_indexer_lambda.download_datafile):
-        pdq_data_file = s3_client.get_object(
-            Bucket=THREAT_EXCHANGE_DATA_BUCKET_NAME, Key=THREAT_EXCHANGE_PDQ_DATA_KEY
-        )
-
-    with metrics.timer(metrics.names.pdq_indexer_lambda.parse_datafile):
-        pdq_data_reader = csv.DictReader(
-            codecs.getreader("utf-8")(pdq_data_file["Body"]),
-            fieldnames=PDQ_DATA_FILE_COLUMNS,
-        )
-        pdq_data = [
-            (
-                row["hash"],
-                # Also add hash to metadata for easy look up on match
-                {
-                    "id": int(row["id"]),
-                    "hash": row["hash"],
-                    "source": "te",  # default for now to make downstream easier to generalize
-                },
-            )
-            for row in pdq_data_reader
+    with metrics.timer(metrics.names.pdq_indexer_lambda.download_datafiles):
+        s3_bucket_files = s3_client.list_objects_v2(
+            Bucket=THREAT_EXCHANGE_DATA_BUCKET_NAME,
+            Prefix=THREAT_EXCHANGE_DATA_FOLDER,
+        )["Contents"]
+        logger.info("Found " + str(len(s3_bucket_files)) + " Files")
+
+        pdq_data_files = [
+            get_pdq_file(file["Key"])
+            for file in s3_bucket_files
+            if file["Key"].endswith(THREAT_EXCHANGE_PDQ_KEY_SUFFIX)
```

Comment:
Hmm, so we're ending up in a case where we are still using S3 triggers, but now we're triggering it multiple times in a row (since there are now multiple files). Each of those triggers will use ls to compute on all of the files that it's triggering for, thus generating n^2 behavior.

This seems non-ideal in the long term. Is there anything that can be done about this?

Reply:
Agreed this could become an issue. `n`  here is number of collaborations not number of indicators correct? So with <10 collabs I don't anticipate it becoming a huge deal.

The solution I see here is to no longer rely on SNS but have a timed job (like the fetcher) that reads the update times of all the pdq files to see if the index needs to be recalculated. Update time is also available in the prefix query

I dont necessarily think its worth implementing this until it becomes an issue

Reply:
It's c*s where c is the number of collabs and s is the number of signal types.

I'm fine punting, please add a TODO somewhere

Reply:
Hmm my math is different (or were solving different problems). I calculate the number of files parsed sa `c^2 * s` currently and this could be optimized to `c * s`. So the difference is a factor of `c < 10` Maybe we're saying the same thing?

Either way, added a TODO on the `was_pdq_data_updated` function